// game_render/js/utils/PlayerManager.js
import constant from './Constant.js';
import Character from '../object/Character.js';

export default class PlayerManager {
    constructor(context, team, dataPlayer) {
        this.team = team;

        this.bomberman = context.add.existing(new Character(context,
            (dataPlayer.bomber.x*16+8 + constant.ORIGIN_X) * constant.SCALE,
            (dataPlayer.bomber.y*16+8 + constant.ORIGIN_Y) * constant.SCALE,
            'bomberman_' + team));
        this.monster = context.add.existing(new Character(context,
            (dataPlayer.enemy.x*16+8 + constant.ORIGIN_X) * constant.SCALE,
            (dataPlayer.enemy.y*16+8 + constant.ORIGIN_Y) * constant.SCALE,
            'monster_' + team));

        this.drawText(context, team);
        this.setData(context, dataPlayer);
    }

    endDrawOrWin() {
        this.bomberman.playStandAnimation();
        this.monster.playStandAnimation();
    }

    endLost() {
        this.bomberman.playDeadAnimation();
        this.monster.playDeadAnimation();
    }

    setData(context, dataPlayer) {
        this.score.setText(dataPlayer.score);
        this.rollers.setText(dataPlayer.rollers);
        this.bombs.setText(dataPlayer.bombs);
        this.flames.setText(dataPlayer.flames);
        this.paws.setText(dataPlayer.paws);

        this.bomberman.setNewXY(
            context,
            (dataPlayer.bomber.x*16 + 8 + constant.ORIGIN_X) * constant.SCALE,
            (dataPlayer.bomber.y*16 + 8 + constant.ORIGIN_Y) * constant.SCALE);
        this.monster.setNewXY(
            context,
            (dataPlayer.enemy.x*16 + 8 + constant.ORIGIN_X) * constant.SCALE,
            (dataPlayer.enemy.y*16 + 8 + constant.ORIGIN_Y) * constant.SCALE);
    }

    drawText(context, team) {
        let image;
        if (team === 1) {
            image = context.add.image(50 * (constant.SCALE / 2.5), 40 * (constant.SCALE / 2.5), 'star_icon');
            image.setScale(constant.SCALE / 2.5, constant.SCALE / 2.5);
            this.score = context.add.text(70 * (constant.SCALE / 2.5), 26 * (constant.SCALE / 2.5), '200', {
                fontFamily: 'Ani',
                fontSize: 30 * (constant.SCALE / 2.5),
                color: '#ffffff'
            }).setStroke(0x111111, 3);

            image = context.add.image(180 * (constant.SCALE / 2.5), 20 * (constant.SCALE / 2.5), 'bomb_icon');
            image.setScale(constant.SCALE / 2.5, constant.SCALE / 2.5);
            this.bombs = context.add.text(205 * (constant.SCALE / 2.5), 6 * (constant.SCALE / 2.5), '2', {
                fontFamily: 'Ani',
                fontSize: 30 * (constant.SCALE / 2.5),
                color: '#ffffff'
            }).setStroke(0x111111, 3);

            image = context.add.image(300 * (constant.SCALE / 2.5), 23 * (constant.SCALE / 2.5), 'roller_icon');
            image.setScale(constant.SCALE / 2.5, constant.SCALE / 2.5);
            this.rollers = context.add.text(325 * (constant.SCALE / 2.5), 6 * (constant.SCALE / 2.5), '2', {
                fontFamily: 'Ani',
                fontSize: 30 * (constant.SCALE / 2.5),
                color: '#ffffff'
            }).setStroke(0x111111, 3);

            image = context.add.image(175 * (constant.SCALE / 2.5), 60 * (constant.SCALE / 2.5), 'flame_icon');
            image.setScale(constant.SCALE / 2.5, constant.SCALE / 2.5);
            this.flames = context.add.text(205 * (constant.SCALE / 2.5), 46 * (constant.SCALE / 2.5), '2', {
                fontFamily: 'Ani',
                fontSize: 30 * (constant.SCALE / 2.5),
                color: '#ffffff'
            }).setStroke(0x111111, 3);

            image = context.add.image(300 * (constant.SCALE / 2.5), 60 * (constant.SCALE / 2.5), 'paw_icon');
            image.setScale(constant.SCALE / 2.5, constant.SCALE / 2.5);
            this.paws = context.add.text(325 * (constant.SCALE / 2.5), 46 * (constant.SCALE / 2.5), '2', {
                fontFamily: 'Ani',
                fontSize: 30 * (constant.SCALE / 2.5),
                color: '#ffffff'
            }).setStroke(0x111111, 3);

            image = context.add.image(450 * (constant.SCALE / 2.5), 40 * (constant.SCALE / 2.5), 'flag_1');
            image.setScale(constant.SCALE / 2.5, constant.SCALE / 2.5);
        } else {
            image = context.add.image(555 * (constant.SCALE / 2.5), 40 * (constant.SCALE / 2.5), 'flag_2');
            image.setScale(constant.SCALE / 2.5, constant.SCALE / 2.5);

            image = context.add.image(650 * (constant.SCALE / 2.5), 23 * (constant.SCALE / 2.5), 'roller_icon');
            image.setScale(constant.SCALE / 2.5, constant.SCALE / 2.5);
            this.rollers = context.add.text(675 * (constant.SCALE / 2.5), 6 * (constant.SCALE / 2.5), '2', {
                fontFamily: 'Ani',
                fontSize: 30 * (constant.SCALE / 2.5),
                color: '#ffffff'
            }).setStroke(0x111111, 3);

            image = context.add.image(770 * (constant.SCALE / 2.5), 23 * (constant.SCALE / 2.5), 'bomb_icon');
            image.setScale(constant.SCALE / 2.5, constant.SCALE / 2.5);
            this.bombs = context.add.text(795 * (constant.SCALE / 2.5), 6 * (constant.SCALE / 2.5), '2', {
                fontFamily: 'Ani',
                fontSize: 30 * (constant.SCALE / 2.5),
                color: '#ffffff'
            }).setStroke(0x111111, 3);

            image = context.add.image(650 * (constant.SCALE / 2.5), 60 * (constant.SCALE / 2.5), 'paw_icon');
            image.setScale(constant.SCALE / 2.5, constant.SCALE / 2.5);
            this.paws = context.add.text(675 * (constant.SCALE / 2.5), 46 * (constant.SCALE / 2.5), '2', {
                fontFamily: 'Ani',
                fontSize: 30 * (constant.SCALE / 2.5),
                color: '#ffffff'
            }).setStroke(0x111111, 3);

            image = context.add.image(765 * (constant.SCALE / 2.5), 60 * (constant.SCALE / 2.5), 'flame_icon');
            image.setScale(constant.SCALE / 2.5, constant.SCALE / 2.5);
            this.flames = context.add.text(795 * (constant.SCALE / 2.5), 46 * (constant.SCALE / 2.5), '2', {
                fontFamily: 'Ani',
                fontSize: 30 * (constant.SCALE / 2.5),
                color: '#ffffff'
            }).setStroke(0x111111, 3);

            image = context.add.image(910 * (constant.SCALE / 2.5), 40 * (constant.SCALE / 2.5), 'star_icon');
            image.setScale(constant.SCALE / 2.5, constant.SCALE / 2.5);
            this.score = context.add.text(930 * (constant.SCALE / 2.5), 26 * (constant.SCALE / 2.5), '200', {
                fontFamily: 'Ani',
                fontSize: 30 * (constant.SCALE / 2.5),
                color: '#ffffff'
            }).setStroke(0x111111, 3);
        }
    }
}
