// game_render/js/scenes/LoadingScene.js
import Phaser from '../phaser.js';

import start from '../../assets/scene/start.jpg';
import win from '../../assets/scene/win.png';
import lose from '../../assets/scene/lose.png';
import draw from '../../assets/scene/draw.png';

import mine from '../../assets/map/mine.png';
import grass from '../../assets/map/grass.png';
import ground from '../../assets/map/ground.png';
import wall from '../../assets/map/wall.png';

import powerup_bombs from '../../assets/item/powerup_bombs.png';
import powerup_flames from '../../assets/item/powerup_flames.png';
import powerup_speed from '../../assets/item/powerup_speed.png';
import powerup_stars from '../../assets/item/powerup_stars.png';

import star_icon from '../../assets/info/star.png';
import bomb_icon from '../../assets/info/bomb.png';
import brick_icon from '../../assets/info/brick.png';
import flame_icon from '../../assets/info/flame.png';
import roller_icon from '../../assets/info/roller.png';
import paw_icon from '../../assets/info/paw.png';
import flag_1 from '../../assets/info/flag_1.png';
import flag_2 from '../../assets/info/flag_2.png';

import bomberman_1_1 from '../../assets/character/bomberman_1/bomberman_1_1.png';
import bomberman_1_2 from '../../assets/character/bomberman_1/bomberman_1_2.png';
import bomberman_1_3 from '../../assets/character/bomberman_1/bomberman_1_3.png';
import bomberman_1_4 from '../../assets/character/bomberman_1/bomberman_1_4.png';
import bomberman_1_5 from '../../assets/character/bomberman_1/bomberman_1_5.png';
import bomberman_1_6 from '../../assets/character/bomberman_1/bomberman_1_6.png';
import bomberman_1_7 from '../../assets/character/bomberman_1/bomberman_1_7.png';
import bomberman_1_8 from '../../assets/character/bomberman_1/bomberman_1_8.png';
import bomberman_1_9 from '../../assets/character/bomberman_1/bomberman_1_9.png';
import bomberman_1_10 from '../../assets/character/bomberman_1/bomberman_1_10.png';
import bomberman_1_11 from '../../assets/character/bomberman_1/bomberman_1_11.png';
import bomberman_1_12 from '../../assets/character/bomberman_1/bomberman_1_12.png';
import bomberman_1_dead_1 from '../../assets/character/bomberman_1_dead/bomberman_1_dead_1.png';
import bomberman_1_dead_2 from '../../assets/character/bomberman_1_dead/bomberman_1_dead_2.png';
import bomberman_1_dead_3 from '../../assets/character/bomberman_1_dead/bomberman_1_dead_3.png';

import bomberman_2_1 from '../../assets/character/bomberman_2/bomberman_2_1.png';
import bomberman_2_2 from '../../assets/character/bomberman_2/bomberman_2_2.png';
import bomberman_2_3 from '../../assets/character/bomberman_2/bomberman_2_3.png';
import bomberman_2_4 from '../../assets/character/bomberman_2/bomberman_2_4.png';
import bomberman_2_5 from '../../assets/character/bomberman_2/bomberman_2_5.png';
import bomberman_2_6 from '../../assets/character/bomberman_2/bomberman_2_6.png';
import bomberman_2_7 from '../../assets/character/bomberman_2/bomberman_2_7.png';
import bomberman_2_8 from '../../assets/character/bomberman_2/bomberman_2_8.png';
import bomberman_2_9 from '../../assets/character/bomberman_2/bomberman_2_9.png';
import bomberman_2_10 from '../../assets/character/bomberman_2/bomberman_2_10.png';
import bomberman_2_11 from '../../assets/character/bomberman_2/bomberman_2_11.png';
import bomberman_2_12 from '../../assets/character/bomberman_2/bomberman_2_12.png';
import bomberman_2_dead_1 from '../../assets/character/bomberman_2_dead/bomberman_2_dead_1.png';
import bomberman_2_dead_2 from '../../assets/character/bomberman_2_dead/bomberman_2_dead_2.png';
import bomberman_2_dead_3 from '../../assets/character/bomberman_2_dead/bomberman_2_dead_3.png';

import monster_1_1 from '../../assets/character/monster_1/monster_1_1.png';
import monster_1_2 from '../../assets/character/monster_1/monster_1_2.png';
import monster_1_3 from '../../assets/character/monster_1/monster_1_3.png';
import monster_1_4 from '../../assets/character/monster_1/monster_1_4.png';
import monster_1_5 from '../../assets/character/monster_1/monster_1_5.png';
import monster_1_6 from '../../assets/character/monster_1/monster_1_6.png';
import monster_1_7 from '../../assets/character/monster_1/monster_1_7.png';

import monster_2_1 from '../../assets/character/monster_2/monster_2_1.png';
import monster_2_2 from '../../assets/character/monster_2/monster_2_2.png';
import monster_2_3 from '../../assets/character/monster_2/monster_2_3.png';
import monster_2_4 from '../../assets/character/monster_2/monster_2_4.png';
import monster_2_5 from '../../assets/character/monster_2/monster_2_5.png';
import monster_2_6 from '../../assets/character/monster_2/monster_2_6.png';
import monster_2_7 from '../../assets/character/monster_2/monster_2_7.png';

import bomb_1 from '../../assets/object/bomb/bomb_1.png';
import bomb_2 from '../../assets/object/bomb/bomb_2.png';
import bomb_3 from '../../assets/object/bomb/bomb_3.png';
import bomb_4 from '../../assets/object/bomb/bomb_4.png';
import bomb_5 from '../../assets/object/bomb/bomb_5.png';
import bomb_6 from '../../assets/object/bomb/bomb_6.png';
import bomb_7 from '../../assets/object/bomb/bomb_7.png';
import bomb_8 from '../../assets/object/bomb/bomb_8.png';
import bomb_9 from '../../assets/object/bomb/bomb_9.png';
import bomb_10 from '../../assets/object/bomb/bomb_10.png';
import bomb_11 from '../../assets/object/bomb/bomb_11.png';
import bomb_12 from '../../assets/object/bomb/bomb_12.png';
import bomb_13 from '../../assets/object/bomb/bomb_13.png';
import bomb_14 from '../../assets/object/bomb/bomb_14.png';
import bomb_15 from '../../assets/object/bomb/bomb_15.png';

import brick_1 from '../../assets/object/brick/brick_1.png';
import brick_2 from '../../assets/object/brick/brick_2.png';
import brick_3 from '../../assets/object/brick/brick_3.png';
import brick_4 from '../../assets/object/brick/brick_4.png';
import brick_5 from '../../assets/object/brick/brick_5.png';
import brick_6 from '../../assets/object/brick/brick_6.png';
import brick_7 from '../../assets/object/brick/brick_7.png';
import brick_8 from '../../assets/object/brick/brick_8.png';

import explosion_1 from '../../assets/object/explosion/explosion_1.png';
import explosion_2 from '../../assets/object/explosion/explosion_2.png';
import explosion_3 from '../../assets/object/explosion/explosion_3.png';
import explosion_4 from '../../assets/object/explosion/explosion_4.png';
import explosion_5 from '../../assets/object/explosion/explosion_5.png';
import explosion_6 from '../../assets/object/explosion/explosion_6.png';
import explosion_7 from '../../assets/object/explosion/explosion_7.png';
import explosion_8 from '../../assets/object/explosion/explosion_8.png';
import explosion_9 from '../../assets/object/explosion/explosion_9.png';
import explosion_10 from '../../assets/object/explosion/explosion_10.png';
import explosion_11 from '../../assets/object/explosion/explosion_11.png';
import explosion_12 from '../../assets/object/explosion/explosion_12.png';
import explosion_13 from '../../assets/object/explosion/explosion_13.png';
import explosion_14 from '../../assets/object/explosion/explosion_14.png';
import explosion_15 from '../../assets/object/explosion/explosion_15.png';
import explosion_16 from '../../assets/object/explosion/explosion_16.png';
import explosion_17 from '../../assets/object/explosion/explosion_17.png';
import explosion_18 from '../../assets/object/explosion/explosion_18.png';
import explosion_19 from '../../assets/object/explosion/explosion_19.png';
import explosion_20 from '../../assets/object/explosion/explosion_20.png';
import explosion_21 from '../../assets/object/explosion/explosion_21.png';

export default class LoadingScene extends Phaser.Scene {

  constructor() {
    super('LoadingScene');
  }

  preload() {
    this.load.image('startScene', start);
    this.load.image('win', win);
    this.load.image('lose', lose);
    this.load.image('draw', draw);

    this.textures.addBase64('mine', mine);
    this.textures.addBase64('grass', grass);
    this.textures.addBase64('ground', ground);
    this.textures.addBase64('wall', wall);

    let arr;

    arr = [bomberman_1_1, bomberman_1_2, bomberman_1_3, bomberman_1_4, bomberman_1_5, bomberman_1_6,
      bomberman_1_7, bomberman_1_8, bomberman_1_9, bomberman_1_10, bomberman_1_11, bomberman_1_12];
    for (let i = 1; i <= arr.length; i++)
      this.textures.addBase64('bomberman_1_' + i, arr[i - 1]);

    arr = [bomberman_1_dead_1, bomberman_1_dead_2, bomberman_1_dead_3];
    for (let i = 1; i <= arr.length; i++)
      this.textures.addBase64('bomberman_1_dead_' + i, arr[i - 1]);

    arr = [bomberman_2_1, bomberman_2_2, bomberman_2_3, bomberman_2_4, bomberman_2_5, bomberman_2_6,
      bomberman_2_7, bomberman_2_8, bomberman_2_9, bomberman_2_10, bomberman_2_11, bomberman_2_12];
    for (let i = 1; i <= arr.length; i++)
      this.textures.addBase64('bomberman_2_' + i, arr[i - 1]);

    arr = [bomberman_2_dead_1, bomberman_2_dead_2, bomberman_2_dead_3];
    for (let i = 1; i <= arr.length; i++)
      this.textures.addBase64('bomberman_2_dead_' + i, arr[i - 1]);

    arr = [monster_1_1, monster_1_2, monster_1_3, monster_1_4, monster_1_5, monster_1_6, monster_1_7];
    for (let i = 1; i <= arr.length; i++)
      this.textures.addBase64('monster_1_' + i, arr[i - 1]);

    arr = [monster_2_1, monster_2_2, monster_2_3, monster_2_4, monster_2_5, monster_2_6, monster_2_7];
    for (let i = 1; i <= arr.length; i++)
      this.textures.addBase64('monster_2_' + i, arr[i - 1]);

    arr = [bomb_1, bomb_2, bomb_3, bomb_4, bomb_5, bomb_6, bomb_7, bomb_8, bomb_9,
      bomb_10, bomb_11, bomb_12, bomb_13, bomb_14, bomb_15];
    for (let i = 1; i <= arr.length; i++)
      this.textures.addBase64('bomb_' + i, arr[i - 1]);

    arr = [brick_1, brick_2, brick_3, brick_4, brick_5, brick_6, brick_7, brick_8];
    for (let i = 1; i <= arr.length; i++)
      this.textures.addBase64('brick_' + i, arr[i - 1]);

    arr = [explosion_1, explosion_2, explosion_3, explosion_4, explosion_5, explosion_6, explosion_7,
      explosion_8, explosion_9, explosion_10, explosion_11, explosion_12, explosion_13, explosion_14,
      explosion_15, explosion_16, explosion_17, explosion_18, explosion_19, explosion_20, explosion_21];
    for (let i = 1; i <= arr.length; i++)
      this.textures.addBase64('explosion_' + i, arr[i - 1]);

    this.textures.addBase64('powerup_bombs', powerup_bombs);
    this.textures.addBase64('powerup_flames', powerup_flames);
    this.textures.addBase64('powerup_speed', powerup_speed);
    this.textures.addBase64('powerup_stars', powerup_stars);

    this.textures.addBase64('star_icon', star_icon);
    this.textures.addBase64('bomb_icon', bomb_icon);
    this.textures.addBase64('brick_icon', brick_icon);
    this.textures.addBase64('flame_icon', flame_icon);
    this.textures.addBase64('roller_icon', roller_icon);
    this.textures.addBase64('paw_icon', paw_icon);
    this.textures.addBase64('flag_1', flag_1);
    this.textures.addBase64('flag_2', flag_2);
  }

  create() {
    this.createAnimationForBomberman('bomberman_1');
    this.createAnimationForBomberman('bomberman_2');
    this.createAnimationForMonster('monster_1');
    this.createAnimationForMonster('monster_2');

    this.createAnimationForBomb();
    this.createAnimationForBrick();
    this.createAnimationForExplosion();

    this.add.text(0, 0, '0', {fontFamily: 'Ani', fontSize: 1});
    this.add.text(0, 0, '0', {fontFamily: 'allamni', fontSize: 1});
  }

  createAnimationForBomberman(name) {
    ['_up', '_right', '_down', '_left'].forEach(function (element, index) {
      this.anims.create({
        key: name + element,
        frames: [{key: name + '_' + (1 + index)}, {key: name + '_' + (5 + index)}, {key: name + '_' + (9 + index)}],
        frameRate: 5,
        repeat: -1
      });
    }.bind(this));

    this.anims.create({
      key: name + '_stand',
      frames: [{key: name + '_3'}],
      frameRate: 5
    });

    this.anims.create({
      key: name + '_dead',
      frames: [{key: name + '_dead_1'}, {key: name + '_dead_2'}, {key: name + '_dead_3'}],
      frameRate: 5,
    });
  }

  createAnimationForMonster(name) {
    ['_left', '_right', '_up', '_down'].forEach(function (element, index) {
      this.anims.create({
        key: name + element,
        frames: (index < 2) ?
          [{key: name + '_' + (1 + index)}, {key: name + '_' + (3 + index)}, {key: name + '_' + (5 + index)}] :
          [{key: name + '_1'}, {key: name + '_2'}, {key: name + '_3'}],
        frameRate: 5,
        repeat: -1
      });
    }.bind(this));

    this.anims.create({
      key: name + '_stand',
      frames: [{key: name + '_3'}],
      frameRate: 5
    });

    this.anims.create({
      key: name + '_dead',
      frames: [{key: name + '_1'}, {key: name + '_2'}, {key: name + '_7'}],
      frameRate: 5
    });
  }

  createAnimationForBomb() {
    this.anims.create({
      key: 'bomb_animation',
      frames: [{key: 'bomb_1'}, {key: 'bomb_2'}, {key: 'bomb_3'}, {key: 'bomb_4'}, {key: 'bomb_5'},
        {key: 'bomb_6'}, {key: 'bomb_7'}, {key: 'bomb_8'}, {key: 'bomb_9'}, {key: 'bomb_10'}, {key: 'bomb_11'},
        {key: 'bomb_12'}, {key: 'bomb_12'}, {key: 'bomb_13'}, {key: 'bomb_14'}, {key: 'bomb_15'}],
      frameRate: 7,
      repeat: -1
    });
  }

  createAnimationForBrick() {
    this.anims.create({
      key: 'brick_animation',
      frames: [{key: 'brick_1'}, {key: 'brick_2'}, {key: 'brick_3'}, {key: 'brick_4'},
        {key: 'brick_5'}, {key: 'brick_6'}, {key: 'brick_7'}, {key: 'brick_8'}],
      frameRate: 15
    });
  }

  createAnimationForExplosion() {
    ['_center', '_horizontal', '_left', '_right', '_vertical', '_down', '_up'].forEach(function (element, index) {
      this.anims.create({
        key: 'explosion' + element,
        frames: [{key: 'explosion_' + (1+index)}, {key: 'explosion_' + (8+index)}, {key: 'explosion_' + (15+index)}],
        frameRate: 15
      });
    }.bind(this));
  }
}
