// game_render/js/scenes/StartScene.js
import Phaser from '../phaser.js';

import constant from '../utils/Constant.js';

export default class StartScene extends Phaser.Scene {

    constructor() {
        super('StartScene');
    }

    init(dataGame) {
        console.log('StartScene', dataGame);
        this.dataGame = dataGame;
    }

    create() {
        let name1 = this.dataGame.player_1;
        let name2 = this.dataGame.player_2;

        if (name1.length > 18)
            name1 = name1.substring(0, 15) + "...";
        if (name2.length > 18)
            name2 = name2.substring(0, 15) + "...";

        this.add.image(constant.MAP_WIDTH / 2 * constant.SCALE, constant.MAP_HEIGHT / 2 * constant.SCALE, 'startScene');

        this.name_1 = this.add.text(-300 * (constant.SCALE / 2.5), constant.MAP_HEIGHT * constant.SCALE * 0.25, name1, {
            fontSize: 40 * (constant.SCALE / 2.5),
            fontStyle: 'bold',
            fontFamily: 'allamni',
            color: '#e32e1b'
        });

        this.name_2 = this.add.text(constant.MAP_WIDTH * constant.SCALE + 300 * (constant.SCALE / 2.5), constant.MAP_HEIGHT * constant.SCALE * 0.7, name2, {
            fontSize: 40 * (constant.SCALE / 2.5),
            fontStyle: 'bold',
            fontFamily: 'allamni',
            color: '#1a2738'
        });

        this.tweens.add({
            targets: this.name_1,
            x: constant.MAP_WIDTH * constant.SCALE * 0.42 - this.name_1.width,
            y: constant.MAP_HEIGHT * constant.SCALE * 0.25,
            duration: 1000,
            ease: 'Back.easeOut'
        });

        this.tweens.add({
            targets: this.name_2,
            x: constant.MAP_WIDTH * constant.SCALE * 0.6,
            y: constant.MAP_HEIGHT * constant.SCALE * 0.7,
            duration: 1000,
            ease: 'Back.easeOut'
        });

        this.time.delayedCall(1500, function () {
            this.scene.start('GameScene', this.dataGame);
        }, [], this);
    }
}
