// game_render/js/object/Character.js
import Phaser from '../phaser.js';

import constant from "../utils/Constant.js";

export default class Character extends Phaser.GameObjects.Sprite {

    constructor (context, x, y, _name) {
        super(context, x, y, _name);

        this.name = _name;
        this.oldX = x;
        this.oldY = y;

        this.setDepth(2);
        this.setScale(constant.SCALE, constant.SCALE);

        this.anims.play(this.name + '_stand', true);
    }

    preUpdate(time, delta) {
        super.preUpdate(time, delta);
    }

    setNewXY(context, x, y) {
        if (x === this.oldX && y === this.oldY)
            this.anims.play(this.name + '_stand', true);
        if (x < this.oldX && y === this.oldY)
            this.anims.play(this.name + '_left', true);
        if (x > this.oldX && y === this.oldY)
            this.anims.play(this.name + '_right', true);
        if (x === this.oldX && y < this.oldY)
            this.anims.play(this.name + '_up', true);
        if (x === this.oldX && y > this.oldY)
            this.anims.play(this.name + '_down', true);

        this.oldX = this.x;
        this.oldY = this.y;

        context.tweens.add({
            targets: this,
            x: x,
            y: y,
            duration: constant.TIMEGAP
        });
    }

    playStandAnimation() {
        this.anims.play(this.name + '_stand', true);
    }

    playDeadAnimation() {
        this.anims.play(this.name + '_dead', true);
    }
}
